### DISCUSSION TITLE
Cariboulite on Rasperry Pi Zero 2 W

### DISCUSSION THREAD

**Discussion by hikariuk** [cariboulabs/cariboulite#206] [Q&A]
Does anyone have any tricks for getting the cariboulite working? I've tried it with Buster and Bookworm, both 32-bit and 64-bit images and get the same result when trying to read from it:

```
C0 D4 FF 1E C0 00 E0 18  C0 00 C0 00 C0 00 C0 00  |  ................
04-30 18:43:13.174  1428  1428 E CARIBOULITE Radio cariboulite_radio_read_samples@cariboulite_radio.c:1276 SMI data synchronization failed
04-30 18:43:13.174  1428  1428 E CARIBOULITE Main main@cariboulite_util.c:345 Samples read operation failed. Quiting...
```

repeated ad nauseam.

cariboulite_test_app and SoapSDRUtil find the device without an issue, and the flashing also works.

```
Board Information (HAT)
        Category name: hat
        Product name: CaribouLite RPI Hat
        Product ID: 0x0002, Numeric: 2
        Product Version: 0x0001, Numeric: 1
        Product UUID: 05c383f0-5e51-4d57-82d4-2058ebeea816, Numeric serial: 0xD9468D5F
        Vendor: CaribouLabs LTD
04-30 18:48:37.898  2370  2370 D AT86RF215_Main at86rf215_print_version@at86rf215.c:290 MODEM Version: AT86RF215IQ (without basebands), version: 03
        Product Type: CaribouLite ISM
Library Versions:
        (Major, Minor, Rev): (1, 2, 0)

######################################################
##     Soapy SDR -- the SDR abstraction library     ##
######################################################

[INFO] SoapyCaribouliteSession, sessionCount: 0
04-30 18:47:21.642  2068  2068 I FPGA caribou_fpga_program_to_fpga@caribou_fpga.c:210 FPGA already operational - not programming (use 'force_prog=true' to force update)
Printing 'findCariboulite' Request:
Found device 0
  channel = S1G
  device_id = 0
  driver = Cariboulite
  label = CaribouLite S1G[b28d1abe]
  name = CaribouLite RPI Hat
  serial = b28d1abe
  uuid = 05c383f0-5e51-4d57-82d4-2058ebeea816
  vendor = CaribouLabs LTD
  version = 0x0001

Found device 1
  channel = HiF
  device_id = 1
  driver = Cariboulite
  label = CaribouLite HiF[b28d1abf]
  name = CaribouLite RPI Hat
  serial = b28d1abf
  uuid = 05c383f0-5e51-4d57-82d4-2058ebeea816
  vendor = CaribouLabs LTD
  version = 0x0001

```

**Comment by thomasbruninx** [comment]
Hi,

I've got the same issue on my RPI4. Every SDR application I use gives me this error: `CARIBOULITE Radio cariboulite_radio_read_samples@cariboulite_radio.c:1276 SMI data synchronization failed`.

Did you ever figure out a solution by any chance?

Thanks in advance

**Reply by hikariuk** [reply]
Unfortunately not.

**Reply by thomasbruninx** [reply]
No problem, thanks for your fast reply. If I find something in the days to come I'll let you know.

**Reply by Regrit** [reply]
I wonder if this problem is specific to (or more common with) the ISM version. I have CaribouLite ISM and have faced this SMI synchronization problem with both Pi Zero 2 W and Pi 4B regardless of 64/32 bit Raspbian/Raspbian Lite/Dragon OS. The only thing I have successfully accomplished is sending a CW tone using the cariboulite test app.

**Reply by hikariuk** [reply]
I've wondered the same thing; I'm disinclined to buy a full fat 6GHz version to find out though...

**Reply by akira25** [reply]
We got an ISM-Version working, but got this error on the full-fat-version.

**Comment by unixpunk** [comment]
are you seeing /dev/smi?  can you check uname -r?  you might have conflicting options in config.txt, or you're running 6.8 kernel?  its on my list to double-check functionality on zero 2w - i can say that it does not have enough memory to run any graphical interface along with the cariboulite - quite plainly get an out of memory error taking an older, working SD card of DragonOS from a pi4 and trying to run any SDR app on it on a zero 2.  next test planned is to reinstall with a fresh OS, no UI and get SoapyRemote working again, which i did have working on a pi3b+.  Side note, mine is the full 6ghz model; don't have ISM to test.

In case you haven't seen this thread: https://github.com/cariboulabs/cariboulite/issues/200
But the initial workflow posted does work - but looking toward the bottom, it doesn't work on anything running kernel 6.8 because it seems it breaks /dev/smi.  Hope this helps, once I get the chance to test further I will post my results and details.

**Reply by Regrit** [reply]
I'm 19% sure it is a bug in the code base. Perhaps final commits were tested on the full version but not ISM? When I run the cariboulite_test_app, I can print board info (2), program the FPGA (3), pass a self-test (4), even send what sounds like a CW tone (8) captured on my handheld. However, all attempts to receive an I/Q stream fail regardless of Pi hardware or Linux kernel. The error posted at the top of this thread is exactly what I see, i.e. 16 bytes of something then a failure to synchronize error, rinse and repeat.

As I understand it, "synchronization" is required because Linux can randomly interfere with SMI transfer. The function caribou_smi_rx_data_analyze()  tries to detect missing data and reduce audio glitches by interpolating between good samples. Analyze fails when it calls caribou_smi_find_buffer_offset().

Go to the Degobah System. Find the offset, Luke.

It doesn't make a lot of sense to me why this error would be specific to ISM boards. Both versions appear to use the same FPGA firmware and SMI transfer code...

PS - I was able to run CubicSDR, GQRX, etc... locally on the zero 2w, but you must start with a minimal build like Raspbian Lite then build a minimal X system + windows manager. It's a pain for sure. I upgraded to a 4B 4GB thinking it would solve all of my caribou problems. It did not.

**Reply by akira25** [reply]
I experience this on a full-Board too. ISM-Board works fine on the same platform.

**Reply by Regrit** [reply]
That's really interesting. I wonder if one of the pins is very heat sensitive and a few of us damaged this circuit soldering the header?

**Reply by hikariuk** [reply]
I’m wondering if it might just be some luck of the draw issue with one or more component performances, either on the Caribou board, the Pi board, or both…

**Comment by unixpunk** [comment]
Hopefully some here find this helpful: https://github.com/cariboulabs/cariboulite/discussions/234
